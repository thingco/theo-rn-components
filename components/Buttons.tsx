import React from "react";
import { TouchableOpacity, View } from "react-native";
import { BaseText } from "./Base";
import { Icon } from "./Icons";

import s from "../Styles";

/** Buttons used exclusively in headers: */
export const CloseButton = ({ navigation }) => (
  <TouchableOpacity
    onPress={() => {
      navigation.goBack(null);
    }}
  >
    <Icon selectorString="icons/close" width={24} height={24} color="#ffffff" />
  </TouchableOpacity>
);

export const BackButton = ({ navigation }) => (
  <TouchableOpacity
    onPress={() => {
      navigation.goBack(null);
    }}
  >
    <Icon selectorString="icons/back" width={24} height={24} color="#ffffff" />
  </TouchableOpacity>
);

/** Core styles for buttons/button text that are shared between all variants: */
const buttonStyles = [
  s.align_center,
  s.justify_center,
  s.min_button_height,
  s.rounded,
];

const buttonTextStyles = [s.text_centre, s.text_lg, s.base_font_bold];

/** The three different varieties of button styling: */
export const ButtonPrimary = (props) => (
  <TouchableOpacity
    style={[s.bg_light, ...buttonStyles, props.style]}
    onPress={props.onPress}
  >
    <BaseText style={[s.white, ...buttonTextStyles]}>{props.children}</BaseText>
  </TouchableOpacity>
);

export const ButtonSecondary = (props) => (
  <TouchableOpacity
    style={[...buttonStyles, s.border_2, s.border_light, props.style]}
    onPress={props.onPress}
  >
    <BaseText style={[s.white, ...buttonTextStyles]}>{props.children}</BaseText>
  </TouchableOpacity>
);

export const ButtonWarning = (props) => (
  <TouchableOpacity
    style={[s.bg_contrast, ...buttonStyles, props.style]}
    onPress={props.onPress}
  >
    <BaseText style={[s.white, ...buttonTextStyles]}>{props.children}</BaseText>
  </TouchableOpacity>
);
