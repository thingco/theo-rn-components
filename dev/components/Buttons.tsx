import React from "react";
import { TouchableOpacity } from "react-native";
import { Text } from "./Base";

import s from "./Styles";

type ButtonType = "primary" | "secondary" | "danger";

type ButtonProps = {
  buttonType: ButtonType;
  children: import("react").ReactNode;
  disabled?: boolean;
  onPress: () => void;
  style?: Array<import("react-native").TextStyle>;
};

/** Core styles for buttons/button text that are shared between all variants: */
const baseButtonStyles = [
  s.align_center,
  s.justify_center,
  s.min_button_height,
  s.ph_lg,
  s.rounded,
];

const buttonStyles = (type: ButtonType): any[] => {
  switch (type) {
    case "primary":
      return [...baseButtonStyles, s.bg_light];
    case "secondary":
      return [...baseButtonStyles, s.border_2, s.border_light];
    case "danger":
      return [...baseButtonStyles, s.bg_contrast];
    default:
      throw new Error(
        `Incorrect button type specified (${type}) - type must be one of "primary", "secondary" or "danger".`
      );
  }
};

const buttonTextStyles = [s.text_centre, s.text_lg, s.base_font_bold];

const Button: React.FC<ButtonProps> = ({
  buttonType,
  ...props
}: ButtonProps) => (
  <TouchableOpacity
    style={[
      ...buttonStyles(buttonType),
      { opacity: props.disabled ? 0.5 : 1 },
      props.style,
    ]}
    onPress={props.onPress}
  >
    <Text style={[s.white, ...buttonTextStyles]}>{props.children}</Text>
  </TouchableOpacity>
);

export { Button };
