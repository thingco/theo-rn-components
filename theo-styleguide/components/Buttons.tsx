import React from "react";
import { TouchableOpacity } from "react-native";
import { BaseText as Text } from "./Base";

import s from "./Styles";

type ButtonType = "primary" | "secondary" | "danger";

/** Core styles for buttons/button text that are shared between all variants: */
const baseButtonStyles = [
  s.align_center,
  s.justify_center,
  s.min_button_height,
  s.ph_lg,
  s.rounded,
];

const buttonStyles = (type: ButtonType): any[] => {
  switch (type) {
    case "primary":
      return [...baseButtonStyles, s.bg_light];
    case "secondary":
      return [...baseButtonStyles, s.border_2, s.border_light];
    case "danger":
      return [...baseButtonStyles, s.bg_contrast];
    default:
      throw new Error(
        `Incorrect button type specified (${type}) - type must be one of "primary", "secondary" or "danger".`
      );
  }
};

const buttonTextStyles = [s.text_centre, s.text_lg, s.base_font_bold];

export const Button = ({ type, ...props }) => (
  <TouchableOpacity
    style={[
      ...buttonStyles(type),
      { opacity: props.disabled ? 0.5 : 1 },
      props.style,
    ]}
    onPress={props.onPress}
  >
    <Text style={[s.white, ...buttonTextStyles]}>{props.children}</Text>
  </TouchableOpacity>
);
