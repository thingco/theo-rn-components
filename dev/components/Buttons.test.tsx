import React from "react";
import renderer from "react-test-renderer";

import { Button } from "./Buttons";

test("Button renders correctly", () => {
  const tree = renderer
    .create(
      <Button buttonType="primary" onPress={() => void 0}>
        I'm a button!
      </Button>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
