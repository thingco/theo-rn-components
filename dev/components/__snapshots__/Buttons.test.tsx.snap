// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders correctly 1`] = `
<View
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#19988B",
      "borderRadius": 4,
      "justifyContent": "center",
      "minHeight": 55,
      "opacity": 1,
      "paddingHorizontal": 24,
    }
  }
>
  <Text
    style={
      Array [
        Object {
          "fontFamily": "Bariol",
        },
        Object {
          "fontSize": 16,
        },
        Array [
          Object {
            "color": "#FFFFFF",
          },
          Object {
            "textAlign": "center",
          },
          Object {
            "fontSize": 18,
          },
          Object {
            "fontFamily": "BariolBold",
          },
        ],
      ]
    }
  >
    I'm a button!
  </Text>
</View>
`;
